src/record_ts.h

```diff
@@ -0,0 +1,873 @@
+
+        } else if (info.type == ParamType::Input) {
+
+            jitc_log(LogLevel::Debug, " -> param s%u", info.slot);
+
+            if (info.test_uninit && rv.state == RecordVarState::Uninit)
+                jitc_raise("record(): Varaible at slot s%u was read from by "
+                           "operation o%u, but has not yet been initialized! "
+                           "This can happen if the variable was not part of "
+                           "the input but is used by an recorded operation.",
+                           info.slot,
+                           (uint32_t) this->m_recording.operations.size());
+
+            if (info.vtype == VarType::Void)
+                info.vtype = rv.type;
+        }
+
+        this->m_recording.dependencies.push_back(info);
+    }
+    /// Helper function for recording input parameters given the slot.
+    void add_in_param(uint32_t slot, VarType vtype = VarType::Void,
+                      bool test_uninit = true) {
+        ParamInfo info;
+        info.type        = ParamType::Input;
+        info.slot        = slot;
+        info.test_uninit = test_uninit;
+        info.vtype       = vtype;
+        add_param(info);
+    }
+    /// Helper function recording input access given the pointer.
+    void add_in_param(const void *ptr, VarType vtype = VarType::Void,
+                      bool test_uninit = true) {
+        uint32_t slot = this->get_variable(ptr);
+        add_in_param(slot, vtype, test_uninit);
+    }
+    /// Helper function recording an output access, given the slot and \ref
+    /// VarType
+    void add_out_param(uint32_t slot, VarType vtype) {
+        ParamInfo info;
+        info.type  = ParamType::Output;
+        info.slot  = slot;
+        info.vtype = vtype;
+        add_param(info);
+    }
+    /// Helper function recording an output access, given the pointer and \ref
+    /// VarType
+    void add_out_param(const void *ptr, VarType vtype) {
+        RecordVariable rv;
+        uint32_t slot = this->add_variable(ptr, rv);
+        add_out_param(slot, vtype);
+    }
+    /// Helper function recording an output access, given the pointer and the
+    /// uint32_t representation of a \ref VarType
+    void add_out_param(uint32_t slot, uint32_t vtype) {
+        add_out_param(slot, (VarType) vtype);
+    }
+};
+
+void jitc_freeze_start(JitBackend backend, const uint32_t *inputs,
+                       uint32_t n_inputs);
+
+Recording *jitc_freeze_stop(JitBackend backend, const uint32_t *outputs,
+                            uint32_t n_outputs);
+
+void jitc_freeze_abort(JitBackend backend);
+
+void jitc_freeze_destroy(Recording *recording);
+
+bool jitc_freeze_pause(JitBackend backend);
+
+bool jitc_freeze_resume(JitBackend backend);
+
+void jitc_freeze_replay(Recording *recording, const uint32_t *inputs,
+                        uint32_t *outputs);
+
+int jitc_freeze_dry_run(Recording *recording, const uint32_t *inputs,
+                        uint32_t *outputs);
```

Comment:
Maybe move these to the top to make them more noticeable, since `RecordThreadState` is so long.